presto-iceberg/src/main/java/io/prestosql/plugin/iceberg/IcebergMetadata.java

```diff
@@ -612,47 +612,45 @@ public void rollback()
-                .transform(IcebergColumnHandle.class::cast)
-                .intersect(table.getPredicate());
-
-        if (newDomain.isNone()) {
-            return Optional.empty();
-        }
+        // Extract identity partition column source ids common to ALL specs
+        Set<Integer> partitionSourceIds = icebergTable.spec().fields().stream()
+                .filter(field -> field.transform().isIdentity())
+                .filter(field -> icebergTable.specs().values().stream().allMatch(spec -> spec.fields().contains(field)))
+                .map(PartitionField::sourceId)
+                .collect(toImmutableSet());
 
-        if (newDomain.equals(table.getPredicate())) {
-            return Optional.empty();
-        }
+        BiPredicate<IcebergColumnHandle, Domain> isIdentityPartition = (column, domain) -> partitionSourceIds.contains(column.getId());
 
-        org.apache.iceberg.Table icebergTable = getIcebergTable(metastore, hdfsEnvironment, session, table.getSchemaTableName());
+        TupleDomain<IcebergColumnHandle> newEnforcedConstraint = constraint.getSummary()
+                .transform(IcebergColumnHandle.class::cast)
+                .filter(isIdentityPartition)
+                .intersect(table.getEnforcedPredicate());
 
-        List<PartitionField> fields = icebergTable.spec().fields().stream()
-                .filter(field -> field.transform().isIdentity())
-                .collect(toImmutableList());
+        TupleDomain<IcebergColumnHandle> newUnenforcedConstraint = constraint.getSummary()
+                .transform(IcebergColumnHandle.class::cast)
+                .filter(isIdentityPartition.negate())
+                .intersect(table.getPredicate());
 
-        // Ensure partition specs in all manifests contain the identity fields from the predicate
-        if (!icebergTable.specs().values().stream().allMatch(spec -> spec.fields().containsAll(fields))) {
+        if (newEnforcedConstraint.equals(table.getEnforcedPredicate())
+                && newUnenforcedConstraint.equals(table.getPredicate())) {
             return Optional.empty();
         }
 
-        Set<Integer> partitionSourceIds = icebergTable.spec().fields().stream()
-                .filter(field -> field.transform().isIdentity())
-                .map(PartitionField::sourceId)
-                .collect(toImmutableSet());
-
-        BiPredicate<IcebergColumnHandle, Domain> contains = (column, domain) -> partitionSourceIds.contains(column.getId());
-        TupleDomain<ColumnHandle> remainingTupleDomain = newDomain.filter(contains.negate()).transform(ColumnHandle.class::cast);
-        TupleDomain<IcebergColumnHandle> enforcedTupleDomain = newDomain.filter(contains);
+        if (newEnforcedConstraint.isNone()) {
+            // If no tuple satisfies the constraint, no need to do additional filtering in the engine
+            newUnenforcedConstraint = TupleDomain.all();
```

Comment:
This seems wrong to me. Shouldn't we return "none" to the engine?

Reply:
removed the special condition.